File: src/main/java/co/bitshifted/xapps/backstage/ctrl/DeploymentController.java
/*
 * Copyright (c) 2020. Bitshift (http://bitshifted.co)
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package co.bitshifted.xapps.backstage.ctrl;

import co.bitshifted.xapps.backstage.content.ContentStorage;
import co.bitshifted.xapps.backstage.dto.DeploymentStatusDTO;
import co.bitshifted.xapps.backstage.exception.ContentException;
import co.bitshifted.xapps.backstage.exception.DeploymentException;
import co.bitshifted.xapps.backstage.service.DeploymentService;
import co.bitshifted.xapps.backstage.util.BackstageFunctions;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import java.io.IOException;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;

import static co.bitshifted.xapps.backstage.BackstageConstants.DEPLOYMENT_CONTENT_TYPE;
import static co.bitshifted.xapps.backstage.BackstageConstants.DEPLOYMENT_STATUS_HEADER;

/**
 * @author Vladimir Djurovic
 */
@RestController
@RequestMapping("/deployment")
@Slf4j
public class DeploymentController {

	@Autowired
	private DeploymentService deploymentService;
	@Autowired
	private ContentStorage contentStorage;

	@RequestMapping(method = RequestMethod.POST, value = "/{appId}/content", consumes = DEPLOYMENT_CONTENT_TYPE)
	public ResponseEntity<Object> handleDeploymentUpload(HttpServletRequest request, @PathVariable String appId) throws DeploymentException, ContentException {
		String contentType = request.getContentType();
		if (!DEPLOYMENT_CONTENT_TYPE.equals(contentType)) {
			throw new DeploymentException("Invalid content type");
		}
		try {
			var deploymentPackagePath = Path.of(contentStorage.uploadDeploymentPackage(request.getInputStream(), appId));
			boolean valid = deploymentService.validateDeploymentContent(deploymentPackagePath);
			if (!valid) {
				throw new DeploymentException("Invalid or corrupt deployment archive");
			}
			deploymentService.processContent(deploymentPackagePath, appId);
			String statusUrl = BackstageFunctions.generateServerUrl(request, "/deployment/status/" + deploymentPackagePath.getParent().getFileName().toString());
			log.debug("Deployment status URL: {}", statusUrl);
			return ResponseEntity.accepted().header(DEPLOYMENT_STATUS_HEADER, statusUrl).build();
		} catch (IOException ex) {
			throw new DeploymentException(ex);
		}
	}

	@RequestMapping(method = RequestMethod.GET, value="/status/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
	public @ResponseBody
	DeploymentStatusDTO getDeploymentStatus(@PathVariable("id")  String id) {
		return deploymentService.getDeploymentStatus(id).convertToDto();
	}

}

File: src/main/kotlin/co/bitshifted/backstage/controller/DeploymentController.kt
/*
 *
 *  * Copyright (c) 2022  Bitshift D.O.O (http://bitshifted.co)
 *  *
 *  * This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 */

package co.bitshifted.backstage.controller

import co.bitshifted.backstage.BackstageConstants
import co.bitshifted.backstage.dto.DeploymentDTO
import co.bitshifted.backstage.dto.DeploymentStatusDTO
import co.bitshifted.backstage.service.DeploymentService
import co.bitshifted.backstage.util.generateServerUrl
import co.bitshifted.backstage.util.logger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import javax.servlet.http.HttpServletRequest

const val BASE_PATH = "/v1/deployments"

@RestController
@RequestMapping(BASE_PATH)
class DeploymentController(
    @Autowired val deploymentService: DeploymentService) {

    val logger = logger(this)

    @PostMapping
    fun startDeployment(@RequestBody deployment : DeploymentDTO, request : HttpServletRequest) : ResponseEntity<String> {
        logger.debug("Running deployment stage one for application id {}", deployment.applicationId)
        val deploymentId = deploymentService.submitDeployment(deployment)
        val statusUrl = generateServerUrl(request, "$BASE_PATH/$deploymentId")

        return ResponseEntity.accepted().header(BackstageConstants.DEPLOYMENT_STATUS_HEADER, statusUrl).build()
    }

    @GetMapping("/{deploymentId}")
    fun getDeployment(@PathVariable deploymentId : String) : ResponseEntity<DeploymentStatusDTO> {
        val deployment = deploymentService.getDeployment(deploymentId)
        return ResponseEntity.ok(deployment)
    }

    @PutMapping(value = ["/{deploymentId}"], consumes = ["application/zip"])
    fun acceptDeploymentArchive(@PathVariable deploymentId: String, @RequestBody content : ByteArray, request : HttpServletRequest) : ResponseEntity<String> {
        logger.debug("Accepted deployment archive for {}", deploymentId)
        deploymentService.submitDeploymentArchive(deploymentId, content.inputStream())
        val statusUrl = generateServerUrl(request, "$BASE_PATH/$deploymentId")
        return ResponseEntity.accepted().header(BackstageConstants.DEPLOYMENT_STATUS_HEADER, statusUrl).build()
    }
}